modules/vba_documents/app/controllers/vba_documents/internal/v0/upload_complete_controller.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require_dependency 'vba_documents/application_controller'
+require_dependency 'vba_documents/object_store'
+
+module VBADocuments
+  module Internal
+    module V0
+      class UploadCompleteController < ApplicationController
+        skip_before_action(:authenticate)
+        before_action :verify_message
+        before_action :verify_topic_arn
+
+        def create
+          case request.headers['x-amz-sns-message-type']
+          when 'Notification'
+            json_params['Message']['Records'].each do |record|
+              upload_id = record['s3']['object']['key']
+              process_upload(upload_id)
+            end
+          when 'SubscriptionConfirmation'
+            client = Aws::SNS::Client.new
+            client.confirm_subscription(
+              topic_arn: json_params['TopicArn'],
+              token: json_params['Token'],
+              authenticate_on_unsubscribe: 'authenticateOnUnsubscribe'
+            )
+          else
+            raise Common::Exceptions::ParameterMissing, 'x-amz-sns-message-type'
+          end
+
+          head :no_content
+        end
+
+        private
+
+        # Make sure the message we're receiving is from the topic
+        # we're expecting to get messages from
+        def verify_topic_arn
+          unless Settings.vba_documents.sns.topic_arn == json_params['TopicArn']
+            raise Common::Exceptions::ParameterMissing, 'TopicArn'
+          end
+        end
+
+        def verify_message
+          verifier = Aws::SNS::MessageVerifier.new
+          unless verifier.authentic?(read_body)
+            raise Common::Exceptions::MessageAuthenticityError.new(
+              raw_post: read_body,
+              signature: json_params['Signature']
+            )
+          end
+        end
+
+        def json_params
+          @json_body ||= JSON.parse(read_body)
+        end
+
+        def read_body
+          @body ||= request.body.read
+        end
+
+        def process_upload(upload_id)
+          upload = VBADocuments::UploadSubmission.where(status: 'pending').find_by(guid: upload_id)
+          store = VBADocuments::ObjectStore.new
+          raise Common::Exceptions::RecordNotFound, upload_id unless upload && store.bucket.object(upload.guid).exists?
+          Rails.logger.info('VBADocuments: Processing: ' + upload.inspect)
+          VBADocuments::UploadProcessor.perform_async(upload_id)
+          upload.update(status: 'uploaded')
```

Comment:
We are introducing a more-than-theoretical race condition here between this controller and the UploadScanner. Especially since the scanner first looks up all pending uploads and then processes them in a loop. 

I think we should make the scanner only pick items that are older than 5 minutes or something like that, making it more explicitly a fallback mechanism.

Reply:
Should we look at the time the database record was created or the last_modified date of the object in s3? Seems like it should be the s3 object.

Reply:
Agreed, based on the s3 object.